client-spark/spark3/src/main/java/org/apache/spark/shuffle/writer/RssShuffleWriter.java

```diff
@@ -402,6 +416,105 @@ private void checkIfBlocksFailed() {
     }
   }
 
+  private boolean needReAssignShuffleServer(Set<Long> failedBlockIds) {
+    boolean result = true;
+    FailedBlockSendTracker failedBlockTracker = shuffleManager.getBlockIdsFailedSendTracker(taskId);
+    for (Long failedBlockId : failedBlockIds) {
+      result =
+          failedBlockTracker.getFailedBlockStatus(failedBlockId).stream()
+              // todo: more status need reassign
+              .allMatch(
+                  trackBlockStatus -> trackBlockStatus.getStatusCode() == StatusCode.NO_BUFFER);
+    }
+    return result;
+  }
+
+  private void reAssignServerForFailedBlockIds(Set<Long> failedBlockIds) {
+    List<ShuffleBlockInfo> reAssignSeverBlockInfoList = Lists.newArrayList();
+    List<ShuffleBlockInfo> failedBlockInfoList = Lists.newArrayList();
+    for (Long failedBlockId : failedBlockIds) {
+      FailedBlockSendTracker failedBlockTracker =
+          shuffleManager.getBlockIdsFailedSendTracker(taskId);
+      Set<TrackBlockStatus> trackBlockStatusSet =
+          failedBlockTracker.getFailedBlockStatus(failedBlockId);
+
+      Map<ShuffleServerInfo, List<TrackBlockStatus>> faultyServerToPartitions =
+          trackBlockStatusSet.stream()
+              .collect(Collectors.groupingBy(d -> d.getShuffleServerInfo()));
+
+      faultyServerToPartitions.entrySet().stream()
+          .forEach(
+              t -> {
+                Set<String> partitionIds =
+                    t.getValue().stream()
+                        .map(x -> String.valueOf(x.getShuffleBlockInfo().getPartitionId()))
+                        .collect(Collectors.toSet());
+                ShuffleServerInfo dynamicShuffleServer =
+                    shuffleManager.getReassignedFaultyServers().get(t.getKey().getId());
+                if (dynamicShuffleServer == null) {
+                  dynamicShuffleServer =
+                      reAssignFaultyShuffleServer(partitionIds, t.getKey().getId());
+                  if (dynamicShuffleServer == null) {
```

Comment:
throw exception directly? Otherwise the failed blocks will be ignored.

Reply:
good catch.